### DISCUSSION TITLE
Define route method dynamically

### DISCUSSION THREAD

**Discussion by Pitu** [kartikk221/hyper-express#116] [Q&A] [answered]
It has come to my attention that something I used to do with both express and fastify was build my own router and then register the routes dynamically. My router logic navigates through folders looking for TypeScript files that export a `run()` method and 3 variables called `url`, `method` and `middlewares`.

The following snippet is a close resemblance of how I expected it to work on HyperExpress.

```ts
// src/routes/version.ts

import type { Response } from 'hyper-express';

export const url = '/version';
export const method = 'GET';

export const run = (res: Response) => {
	res.json({
		version: process.env.npm_package_version
	});
};
```

Now as for my router controller I iterate through files, do some transformations and eventually register said routes with:

```ts
// src/structures/router.ts
// server is a HyperExpress.Server instance

server.any(
	route.url,
	{
		middlewares
	},
	(req: Request, res: Response) => route.run(req, res)
);
```

Now the problem with this is that I'm registering routes as `any` which is not ideal in the least. 
I could manually register them using `server.get() / post() / etc` but trying to do it dynamically it fails:

```ts
// src/structures/router.ts
// server is a HyperExpress.Server instance
// route.method is a string that can be get/post/delete/put

server[route.method](
	route.url,
	{
		middlewares
	},
	(req: Request, res: Response) => route.run(req, res)
);
```

This is the output

```
[23:13:54.911] ERROR (21818): server[route.method] is not a function
    err: {
      "type": "TypeError",
      "message": "server[route.method] is not a function",
      "stack":
          TypeError: server[route.method] is not a function
              at Object.load (/home/dev/projects/chibisafe/src/api/structures/routes.ts:46:25)
              at async start (/home/dev/projects/chibisafe/src/api/main.ts:46:2)
    }
```

Are there any other ways to dynamically assign the method to a route since this attempt fails?

PS: Is there a way to print all the routes registered to the WebServer?

**Comment by kartikk221** [comment, marked as answer]
You should likely implement some form a transformer for the creation part of your code that translates the exported constant method from your router files to a callable method on HyperExpress. Most route creation methods on the HyperExpress `Server` and `Router` components are just the lowercase versions of the original HTTP methods for those requests except for some special cases such as `DELETE` being `del()`. So just be sure to go through the documentation and create special translation cases for these and you should be good to go.

For example:
```js
let callable;
switch (route.method) {
    case 'DELETE:
        // Some route creation methods are different from the HTTP methods hence create cases like this for those
        callable = 'del';
        break;
    default:
        // Most route creation methods in HyperExpress are just lowercase forms of the HTTP methods
        callable = route.method.toLowerCase();
        break;
}

server[callable](
	route.url,
	{
		middlewares
	},
	(req: Request, res: Response) => route.run(req, res)
);
```

**Reply by Pitu** [reply]
Hey, thank you very much for this. I believe the issue was that I wasn't lowercasing the method or typescript wasn't sure the type of method because once I assigned it as a string as per your suggestion it started working. Also thanks a lot for the snipper about `DELETE` and `del()`, I missed that.

As a last question, would you know how to print all the registered routes from a `HyperExpress.Server` instance?

**Reply by kartikk221** [reply]
Yeah, you can view the registered routes and middlewares for a given `Server` instance using the `Server.routes` and `Server.middlewares` properties. For routes, they organized as an `Object -> Method[String] -> Route[Route]` while Middlewares are organized as `Object -> Path[String] -> Middleware[Middleware]`.

**Reply by Pitu** [reply]
I appreciate it. Logging the server instance itself didn't enumerate the properties for me and I couldn't find it in the docs as to why this would happen, only saw that a router instance could be logged.

**Reply by kartikk221** [reply]
No, I mean they are getters not just properties. So try doing `console.log(server.routes)` and `console.log(server.middlewares)`.

**Reply by Pitu** [reply]
That worked wonderfully, thanks you!
